[CI] StatelessConcurrentRefreshIT testWaitUntilShouldNotWaitForUpload failing

**Build Scans:**
- [elasticsearch-intake #42236 / part2](https://gradle-enterprise.elastic.co/s/kczyg3lens7qm)
- [elasticsearch-intake #42208 / part2](https://gradle-enterprise.elastic.co/s/3y6pu4hcaxcbs)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.StatelessConcurrentRefreshIT.testWaitUntilShouldNotWaitForUpload" -Dtests.seed=A6D50751BDA5EA01 -Dtests.locale=en-CH -Dtests.timezone=Europe/Moscow -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.StatelessConcurrentRefreshIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testWaitUntilShouldNotWaitForUpload'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: safeAwait: CountDownLatch did not reach zero within the timeout
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testWaitUntilShouldNotWaitForUpload
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 2 failures in test testWaitUntilShouldNotWaitForUpload (0.5% fail rate in 366 executions)
- [main] 2 failures in step part2 (2.7% fail rate in 73 executions)
- [main] 2 failures in pipeline elasticsearch-intake (2.7% fail rate in 73 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -710,9 +710,6 @@ tests:
 - class: org.elasticsearch.xpack.encryption.spi.test.EncryptedDataHandlerProviderSpiIT
   method: testProviderIsDiscoveredAndHandlerIsInvoked
   issue: https://github.com/elastic/elasticsearch/issues/152303
-- class: org.elasticsearch.xpack.stateless.StatelessConcurrentRefreshIT
-  method: testWaitUntilShouldNotWaitForUpload
-  issue: https://github.com/elastic/elasticsearch/issues/152304
 - class: org.elasticsearch.blobcache.common.SparseFileTrackerTests
   method: testThreadSafety
   issue: https://github.com/elastic/elasticsearch/issues/152305

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessConcurrentRefreshIT.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessConcurrentRefreshIT.java
@@ -282,7 +282,7 @@ public class StatelessConcurrentRefreshIT extends AbstractStatelessPluginIntegTe
             + "org.elasticsearch.indices.recovery:debug",
         reason = "ensure detailed shard relocation information"
     )
-    public void testWaitUntilShouldNotWaitForUpload() throws InterruptedException {
+    public void testWaitUntilShouldNotWaitForUpload() throws Exception {
         startMasterOnlyNode();
         final String indexNode = startIndexNode();
         startSearchNode();
@@ -292,11 +292,12 @@ public class StatelessConcurrentRefreshIT extends AbstractStatelessPluginIntegTe
         createIndex(
             indexName,
             indexSettings(1, 1).put(INDEX_TRANSLOG_FLUSH_THRESHOLD_AGE_SETTING.getKey(), "1h")
-                // The test should pass with scheduled refresh disabled. Randomly enable it for closer resemblance to production
-                .put(
-                    IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(),
-                    randomFrom(TimeValue.MINUS_ONE, IndexSettings.STATELESS_MIN_NON_FAST_REFRESH_INTERVAL)
-                )
+                // Scheduled refresh must stay disabled: this test relies on the relocation's forceRefreshes() being the
+                // refresh that notifies the wait_until refresh listener and counts down afterRefreshLatch (via the overridden
+                // 1-arg refresh(String)). A periodic refresh goes through Engine.maybeRefresh() -> the 3-arg refresh(), which
+                // notifies/consumes the listener WITHOUT counting the latch; the relocation would then find no pending listener
+                // (refreshNeeded() == false), never force a refresh, and afterRefreshLatch would never reach zero.
+                .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.MINUS_ONE)
                 .build()
         );
         ensureGreen(indexName);
@@ -313,6 +314,14 @@ public class StatelessConcurrentRefreshIT extends AbstractStatelessPluginIntegTe
         logger.info("--> Starting the bulk indexing thread with wait_until");
         final var indexingThread = new Thread(() -> bulkIndexDocsWithRefresh(indexName, 20, WriteRequest.RefreshPolicy.WAIT_UNTIL));
         indexingThread.start();
+
+        // Establish the happens-before the rest of the test relies on: wait until all docs have been written on the source
+        // primary (20 immediate + 20 immediate + 20 wait_until = 60). By then the wait_until bulk holds its operation permit
+        // and has registered (or is about to register) its refresh listener on the source, so the relocation's forceRefreshes()
+        // is guaranteed to observe a pending listener (refreshNeeded() == true) and fire the latch-counting refresh. Without
+        // this, under load the relocation could drain all permits before the bulk registers, the bulk would reroute to the new
+        // primary, and afterRefreshLatch would never count down.
+        assertBusy(() -> assertThat(testStateless.indexCounter.get(), equalTo(60)));
 
         // 2. Start the relocation and block it right after the pre-flush. The pre-flush creates and uploads generation N.
         // afterRealFlushCompletedBarrierReference is used so that only this real flush (IS_FLUSH_BY_REFRESH=false) can

